Forms: add read/unread state to the UI

enejb opened the pull request:
Fixes FORMS-302

Adds new feedback/read endpoint and the new UI that lets you mark feedback as read and unread. 

<img width="589" height="549" alt="Screenshot 2025-10-02 at 7 38 58 AM" src="https://github.com/user-attachments/assets/1095dff1-4052-4930-81d8-2195f68594d8" />

This new state is shown as a dot. this is due to the current limitation of the data views which prevent the whole row from being bold. 

## Proposed changes:
* Adds a new endpoints for marking feedback responses from being read or unread. 
* Add UI changes so that you see unread feedback responses as read. 
* Add the ability to mark responses as read or unread. 


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
See FORMS-302

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
* Add a form and fill it out. 
* Notice that this response will show up as unread in the forms UI. 
* Notice that you mark previous responses as read and unread.

enejb committed:
Update the feedback class to have is_unread property

enejb committed:
Add new endpoint that lets us marks things as read and unread

enejb committed:
Update the UI that lets us shows us things that are unread and mark them as read once they are viewed

enejb committed:
Add DataView actions

enejb committed:
changelog

enejb committed:
Update the style

edanzer commented on projects/packages/forms/src/dashboard/inbox/style.scss:
```diff
@@ -506,3 +506,11 @@ body.jetpack_page_jetpack-forms-admin {
+	margin-right: 4.5px;
+	margin-left: -12px;
+	vertical-align: middle;
+}
```
The dot color is red here and in testing. It's also quite small. In the PR description, it's blue and a bit larger, which I like more. That may just be taste, and obviously not blocker. 

<img width="1497" height="509" alt="unread-status" src="https://github.com/user-attachments/assets/9b790dff-71ae-451a-92e8-e22452c1f656" />

edanzer reviewed (commented):
This tests fairly well and is an exciting change to see. 

The good: 
- [x] Once this PR is loaded, any new form responses show as unread (with red dot).
- [x] Interacting with unread messages updates status to read (no more dot)
- [x] Selecting items in bulk and clicking the "Mark unread" button successfully marks them unread (with red dot).
- [x] Using the actions dropdown for a single response to mark an item read or unread works.
- [x] Tests pass

Potential improvements: 
- The dot in my testing is small and red. I like the bigger, blue one in the PR description a bit more :). It is a shame we can't otherwise style the un-read item rows more (ie, bolding or background color). 
- Could be done in a follow up, but now that feedback objects have an unread property, I wonder if we could use that to determine the number we show in the notifications bubble in the admin menu for Jetpack and Forms. As is, these are totally unconnected. 
   - Example: Say I have 5 unread messages and I see that button Jetpack and Forms. Soon as I click the inbox, that gets cleared even though I still have unread messages. Also marking message read or unread also doesn't affect that.
- There's a small TS error that needs resolving. 

<img width="1497" height="509" alt="unread-status" src="https://github.com/user-attachments/assets/6ede9f12-0a2b-47bf-9fd7-119db81b61ac" />

enejb committed:
Make things bold

enejb commented:
Thanks for the review @edanzer! 

> should we should the red bubble. 

Yes we should. This part needs updating. ( I think we should switch to a new option ) and go and update the total unread, Everytime we add a new response and when we switch the read state.  Ideally this value would also change as we update the count live on the page. 

I updated design to be indicate that the Response is unread by making the different elements bold. 
It looks like this now. cc @ilonagl. 

<img width="816" height="417" alt="Screenshot 2025-10-03 at 12 21 46 AM" src="https://github.com/user-attachments/assets/99d9021b-f27e-4700-be0d-d792b54699a4" />

simison committed:
Swap icons and add back indicator dot

CGastrell reviewed (dismissed):
Works great! :shipit:

enejb committed:
Update the sidebar counter to match

Replaces legacy unread feedback count logic with new static methods in Contact_Form_Plugin and Feedback classes, storing the count in a new option. Updates backend to recalculate and return the unread count after marking feedback as read/unread. Frontend now updates the admin menu counter dynamically using the new count from the server, improving accuracy and consistency.

simison commented on projects/packages/forms/src/contact-form/class-contact-form-plugin.php:
```diff
@@ -1406,70 +1406,91 @@ public function allow_feedback_rest_api_type( $post_types ) {
+	 */
+	public static function recalculate_unread_count() {
+		$count = Feedback::get_unread_count();
+		update_option( 'feedback_unread_count_v2', $count );
```
We could use `jetpack_` prefix since we're renaming anyway. 🤔

enejb replied:
Good call!

simison replied:
Is there some pre-loading also happening with `jetpack_` prefixed options, some different helper functions, and perhaps they get synced automatically?

simison committed:
Clarify JS comments

simison committed:
Move menu updated to utils

simison committed:
Always output counter badge

enejb committed:
Make the dot red.

enejb committed:
Add mark as read buttons

enejb committed:
Update the option name

simison commented on projects/packages/forms/src/dashboard/inbox/response.js:
```diff
@@ -326,6 +341,28 @@ const InboxResponse = ( {
 			default: // 'publish' (inbox) or any other status
 				return (
 					<>
+						{ response.is_unread && (
```
Should these actions be available to all, inbox/spam/trash views? 🤔

simison reviewed (dismissed):
Testing overall well; tested also combined counts with and without unread form responses:

<img width="180" height="332" alt="Screenshot 2025-10-08 at 17 06 52" src="https://github.com/user-attachments/assets/d3eea34b-f7ad-40f9-a639-e10069de446b" />
<img width="198" height="314" alt="Screenshot 2025-10-08 at 17 13 15" src="https://github.com/user-attachments/assets/d2a22a75-0725-4e80-a8a0-46fa00120c67" />


In a follow-up, let's iterate on +/-:ing sidebar counter optimistically (https://github.com/Automattic/jetpack/pull/45398) instead of having a delay or have counter not work when starting form zero.

Let's also refresh counts when unread posts are moved as-is to trash, or marked as spam. I don't think we should have those in the counter, only inbox-unreads. In this example count is 3, but one unread is in trash:

<img width="821" height="513" alt="Screenshot 2025-10-08 at 17 22 51" src="https://github.com/user-attachments/assets/1929db51-06ea-43bb-aa88-6f0ab4aa5042" />

edanzer reviewed (commented):
Just retested following latest changes. Everything that was working before is working now but now with other nice improvements. 

- [x] I like the bolding for unread items!
- [x] Great to see the syncing between unread items and the notification bubble on the Jetpack/Forms dashboard menu. 
- [x] Single response action buttons to mark unread are working great. 

Improvements: 
* The only small issue I'm seeing is that there can be a delay between when I read an item / mark an item unread and when I see the number in the menu notification bubble change. I assume that's some delay due to need to round trip an endpoint request to update the notification count. Probably something we should improve, but it could easily be handled in a follow up. 
* UPDATE: Can also confirm from other findings, that if you mark an unread item as spam or trash, the number on the dashboard menu doesn't change. Didn't notice this initially because normally I'd read an item before marking it spam or trashing it. 
* Still some failing tests / static analysis issues

enejb committed:
Update the tests to remove untestable code

enejb committed:
Add read unread buttons to each inbox header

enejb committed:
Fix typescript error

enejb committed:
fix phan

enejb committed:
More tests.

enejb committed:
More tests...
